lib/http3/common.c

```diff
@@ -143,26 +148,20 @@ int h2o_quic_send_datagrams(h2o_quic_ctx_t *ctx, quicly_address_t *dest, quicly_
         for (size_t i = 1; i < num_datagrams - 1; ++i)
             assert(datagrams[i].iov_len == datagrams[0].iov_len);
         uint16_t segsize = (uint16_t)datagrams[0].iov_len;
-        cmsg->cmsg_level = SOL_UDP;
-        cmsg->cmsg_type = UDP_SEGMENT;
-        cmsg->cmsg_len = CMSG_LEN(sizeof(segsize));
-        memcpy(CMSG_DATA(cmsg), &segsize, sizeof(segsize));
-        cmsg = (struct cmsghdr *)((char *)cmsg + CMSG_SPACE(sizeof(segsize)));
+        PUSH_CMSG(SOL_UDP, UDP_SEGMENT, segsize);
         using_gso = 1;
     }
 #endif
 
-    /* send datagrams */
-    struct msghdr mess = {
-        .msg_name = &dest->sa,
-        .msg_namelen = quicly_get_socklen(&dest->sa),
-    };
-    if (cmsg != &cmsgbuf.hdr) {
-        mess.msg_control = &cmsgbuf.buf;
-        mess.msg_controllen = (socklen_t)((char *)cmsg - cmsgbuf.buf);
+    /* commit CMSG length */
+    if (cmsg != NULL) {
+        mess.msg_controllen = (socklen_t)((char *)cmsg - (char *)cmsgbuf.buf) + cmsg->cmsg_len;
```

Comment:
I'm not sure if this calculation is correct. man page says

```
Finally, the msg_controllen field of the msghdr should be set to the sum of the
CMSG_SPACE() of the length of all control messages in the buffer.
```

Are there any guarantees that this `cmsg - start + cmsg->cmsg_len` value is equal to that sum of CMSG_SPACE? Even if so, I feel it's more straightforward
to do `controlllen += CMSG_SPACE(sizeof(value));` in PUSH_CMSG and set it to mess.msg_controlllen here.

Reply:
That's a really keen observation. I've found [a good visualization in RFC 3542](https://datatracker.ietf.org/doc/html/rfc3542#page-62) explaining what these values mean, and it does seem that it explicitly permits such structure (quote: _While sending an application may or may not include padding at the end of last ancillary data in msg_controllen and implementations must accept both as valid._).

Though, I agree with you that we have an inconsistency here and that we can do better. Will see what I can do.

Reply:
Yeah, the sentence you quoted explains all 👍